apis/k8ssandra/v1alpha1/k8ssandracluster_types.go

```diff
@@ -195,6 +196,12 @@ type CassandraClusterTemplate struct {
 	// +optional
 	Datacenters []CassandraDatacenterTemplate `json:"datacenters,omitempty"`
 
+	// CassandraTelemetry defines the desired state for telemetry resources in this K8ssandraCluster.
+	// If telemetry configurations are defined, telemetry resources will be deployed to integrate with
+	// a user-provided monitoring solution (at present, only support for Prometheus is available).
+	// +optional
+	CassandraTelemetry *telemetryapi.TelemetrySpec `json:"cassandraTelemetry,omitempty"`
```

Comment:
Can we change the json field name to just `telemetry`? You already have this for Stargate.

Reply:
Yes, because `Telemetry` for Stargate is wrapped in a `StargateTemplate`, which makes the target object clear. Conversely, `CassandraTelemetry` is a member of `CassandraClusterTemplate.Datacenters` of type `CassandraDatacenterTemplate` which also contains members (for e.g.) `Stargate`, `Reaper` and various other non-Cassandra structs.

It is therefore unclear what the telemetry target is if we simply add `Telemetry` as a field here. 

In fact, I think this API design could use some thought, but that isn't a matter for this PR.

Reply:
This along with some other naming issues are certainly subjective, and I agree a matter for this PR. I will start some separate discussions though before 1.0.0.

Reply:
So the naming aspect is indeed something @jsanda just mentioned wanting to generally review the other day, that's one of the areas we wanted to get some feedback from some external to the team users on what makes sense.  So perhaps we should like you both seem to say save that conversation for a different moment.

From the looks of what we have here it seems we're already mixing patterns in this structure

```
type CassandraClusterTemplate struct {
	// Cluster is the name of the cluster. This corresponds to cluster_name in
	// cassandra.yaml.
	// +kubebuilder:validation:MinLength=2
	Cluster string `json:"cluster,omitempty"`

	// SuperuserSecretName allows to override the default super user secret
	SuperuserSecretName string `json:"superuserSecret,omitempty"`

	// ServerImage is the image for the cassandra container. Note that this should be a
	// management-api image. If left empty the operator will choose a default image based
	// on ServerVersion.
	// +optional
	ServerImage string `json:"serverImage,omitempty"`

	// ServerVersion is the Cassandra version.
	// +kubebuilder:validation:Pattern=(3\.11\.\d+)|(4\.0\.\d+)
	ServerVersion string `json:"serverVersion,omitempty"`

	// Resources is the cpu and memory resources for the cassandra container.
	// +optional
	Resources *corev1.ResourceRequirements `json:"resources,omitempty"`

	// SystemLoggerResources is the cpu and memory resources for the server-system-logger
	// container.
	// +optional
	SystemLoggerResources *corev1.ResourceRequirements `json:"systemLoggerResources,omitempty"`

	// CassandraConfig is configuration settings that are applied to cassandra.yaml and
	// jvm-options for 3.11.x or jvm-server-options for 4.x.
	// +optional
	CassandraConfig *CassandraConfig `json:"config,omitempty"`

	// StorageConfig is the persistent storage requirements for each Cassandra pod. This
	// includes everything under /var/lib/cassandra, namely the commit log and data
	// directories.
	// +optional
	StorageConfig *cassdcapi.StorageConfig `json:"storageConfig,omitempty"`

	// Networking enables host networking and configures a NodePort ports.
	// +optional
	Networking *cassdcapi.NetworkingConfig `json:"networking,omitempty"`

	// Racks is a list of named racks. Note that racks are used to create node affinity. //
	// +optional
	Racks []cassdcapi.Rack `json:"racks,omitempty"`

	// Datacenters a list of the DCs in the cluster.
	// +optional
	Datacenters []CassandraDatacenterTemplate `json:"datacenters,omitempty"`

	// CassandraTelemetry defines the desired state for telemetry resources in this K8ssandraCluster.
	// If telemetry configurations are defined, telemetry resources will be deployed to integrate with
	// a user-provided monitoring solution (at present, only support for Prometheus is available).
	// +optional
	CassandraTelemetry *telemetryapi.TelemetrySpec `json:"cassandraTelemetry,omitempty"`

	// MgmtAPIHeap defines the amount of memory devoted to the management
	// api heap.
	// +optional
	MgmtAPIHeap *resource.Quantity `json:"mgmtAPIHeap,omitempty"`
}
```

Since everything is under the `CassandraClusterTemplate` it seems clear to me that everything inside of it is relevant to `Cassandra`, so I'd certainly be inclined to say that calling it `CassandraTelemetry` is redundant and not necessary.  I'd fall on the side of simplicity there.

But, we also have `CassandraConfig`, which doubles up on the naming as well -- I'd personally support a pattern where we don't repeat the naming when the parent already presents it.  So I'd suggest that we simplify this, and should also consider doing the same for `CassandraConfig`.

Reply:
@jdonenine there's two different instances of CassandraTelemetry [here](https://github.com/Miles-Garnsey/k8ssandra-operator/blob/80e661d3db8b3ff20b3938b49f9bcedfb423c7a1/apis/k8ssandra/v1alpha1/k8ssandracluster_types.go#L275) and [here](https://github.com/Miles-Garnsey/k8ssandra-operator/blob/stargate-prometheus-integration/apis/k8ssandra/v1alpha1/k8ssandracluster_types.go#L203). 

Now, the former one on line 275 is in a struct (`CassandraDatacenterTemplate`) which mixes in Reaper and Stargate. The one you've referenced on line 203 is indeed in a pure Cassandra struct (and, very confusingly, resides in `CassandraClusterTemplate`).

So we have a situation where a `CassandraDatacenterTemplate` mixes Reaper and Stargate structs in with Cassandra structs, but `CassandraClusterTemplate` is purely Cassandra related. I certainly wouldn't call the former one `Telemetry` and the latter one `CassandraTelemetry` because they both do exactly the same thing.

The API structure could use a few tweaks in my opinion.  The Stargate API is a lot cleaner and would be a good reference (single type for both cluster + DC config, more aggressive use of inline json/embedded structs). At the least, it would be worth segregating the Cassandra config into it's own api package to clarify the structure.

_(You can see why I didn't want to get into this here...)_